pkg/broker/broker.go

```diff
@@ -116,6 +118,129 @@ func (a AnsibleBroker) Bootstrap() (*BootstrapResponse, error) {
 	return &BootstrapResponse{SpecCount: len(specs), ImageCount: imageCount}, nil
 }
 
+func (a AnsibleBroker) Recover() (string, error) {
+	// At startup we should write a key to etcd.
+	// Then in recovery see if that key exists, which means we are restarting
+	// and need to try to recover.
+
+	// do we have any jobs that wre still running?
+	// get all /state/*/jobs/* == in progress
+	// For each job, check the status of each of their containers to update
+	// their status in case any of them finished.
+
+	recoverStatuses, err := a.dao.FindJobStateByState(apb.StateInProgress)
+	if err != nil {
+		// no jobs or states to recover, this is OK.
+		if client.IsKeyNotFound(err) {
+			a.log.Info("No jobs to recover")
+			return "", nil
+		}
+		return "", err
+	}
+
+	a.log.Debug("Recovery refresh login")
+
+	// need a client to refresh the login
+	client, cerr := apb.NewClient(a.log)
+	if cerr != nil {
+		a.log.Error("recovery failed: could not create client during recovery")
+		return "", cerr
+	}
+
+	cerr = client.RefreshLoginToken(a.clusterConfig)
+	if cerr != nil {
+		a.log.Error("recovery failed: could not refresh the login token")
+		return "", cerr
+	}
+
+	/*
+		if job was in progress we know instanceuuid & token. do we have a podname?
+		if no, job never started
+			restart
+		if yes,
+			did it finish?
+				yes
+					* update status
+					* extractCreds if available
+				no
+					* create a monitoring job to update status
+	*/
+
+	// let's see if we need to recover any of these
+	for _, rs := range recoverStatuses {
+
+		// since we're here we have an in progress job
+		instance, err := a.dao.GetServiceInstance(rs.InstanceId.String())
+
+		// do we have a podname?
+		if rs.State.Podname == "" {
+			// NO, we do not have a podname
+
+			a.log.Info(fmt.Sprintf("No podname. Attempting to restart job: %s", rs.InstanceId.String()))
+
+			si, err := a.dao.GetServiceInstance(rs.InstanceId.String())
```

Comment:
Can you just use the instance already retrieved? or should we wait to retrieve to here? I don't think we need both.